x-pack/plugins/security/server/authentication/authenticator.ts

```diff
@@ -0,0 +1,415 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import {
+  SessionStorageFactory,
+  SessionStorage,
+  KibanaRequest,
+  LoggerFactory,
+  Logger,
+  HttpServiceSetup,
+  ClusterClient,
+} from '../../../../../src/core/server';
+import { ConfigType } from '../config';
+import { getErrorStatusCode } from '../errors';
+
+import {
+  AuthenticationProviderOptions,
+  AuthenticationProviderSpecificOptions,
+  BaseAuthenticationProvider,
+  BasicAuthenticationProvider,
+  KerberosAuthenticationProvider,
+  SAMLAuthenticationProvider,
+  TokenAuthenticationProvider,
+  OIDCAuthenticationProvider,
+  isSAMLRequestQuery,
+} from './providers';
+import { AuthenticationResult } from './authentication_result';
+import { DeauthenticationResult } from './deauthentication_result';
+import { Tokens } from './tokens';
+
+/**
+ * The shape of the session that is actually stored in the cookie.
+ */
+export interface ProviderSession {
+  /**
+   * Name/type of the provider this session belongs to.
+   */
+  provider: string;
+
+  /**
+   * The Unix time in ms when the session should be considered expired. If `null`, session will stay
+   * active until the browser is closed.
+   */
+  expires: number | null;
+
+  /**
+   * Session value that is fed to the authentication provider. The shape is unknown upfront and
+   * entirely determined by the authentication provider that owns the current session.
+   */
+  state: unknown;
+}
+
+/**
+ * The shape of the login attempt.
+ */
+export interface ProviderLoginAttempt {
+  /**
+   * Name/type of the provider this login attempt is targeted for.
+   */
+  provider: string;
+
+  /**
+   * Login attempt can have any form and defined by the specific provider.
+   */
+  value: unknown;
+}
+
+export interface AuthenticatorOptions {
+  config: Pick<ConfigType, 'sessionTimeout' | 'authc'>;
+  basePath: HttpServiceSetup['basePath'];
+  loggers: LoggerFactory;
+  clusterClient: PublicMethodsOf<ClusterClient>;
+  sessionStorageFactory: SessionStorageFactory<ProviderSession>;
+  isSystemAPIRequest: (request: KibanaRequest) => boolean;
+  getServerBaseURL: () => string;
```

Comment:
__note:__ diff#4: instead of passing protocol/hostname/port tripple I just pass a function that generates server base URL. It's required since we get necessary info from the legacy plugin and it's actually shows the intention better.